File: packages/runtime-cs/src/cs/Numeric.cs
// Numeric primitives matching ECMAScript/Lua-backend semantics exactly.
// Ported by transcription from packages/runtime-py/src/py/gltfi_runtime/
// numeric.py (that file's own header documents, function-by-function, every
// way Python's math library disagrees with JS — the reason runtime-py needed
// this module at all). C#'s System.Math is, by contrast, IEEE-754-754
// compliant end-to-end: Math.Sqrt/Log/Log2/Log10/Asin/Acos/Sin/Cos/Tan/Floor/
// Ceiling/Truncate all already return NaN/+-Infinity for out-of-domain or
// non-finite input instead of throwing (verified against the documented
// .NET behavior for each — e.g. Math.Log(0) is NegativeInfinity, Math.Log(-1)
// is NaN, Math.Asin(2) is NaN, Math.Sin(double.PositiveInfinity) is NaN), and
// `/` and `%` on `double` already implement IEEE div-by-zero -> +-Inf/NaN and
// C-fmod-with-NaN-propagating-zero-divisor semantics natively. So most of
// this file's "Safe*" wrappers exist purely for call-site/readability parity
// with the Python/Lua backends' identical-named helpers (every m.* caller
// still routes through Numeric.SafeFoo, not raw Math.Foo, so a future
// contributor scanning this backend sees the same shape as the other two) —
// only ToInt32, SafeRound, and SafePow have any real logic of their own; see
// each one's comment for why.
using System;

namespace GltfiRuntime;

public static class Numeric
{
    public static bool IsNaN(double x) => double.IsNaN(x);

    // Order-independent NaN-propagating min/max (JS's Math.min/Math.max) —
    // .NET's own Math.Min/Math.Max are ALSO order-independent NaN-propagating
    // already (unlike Python's builtin min/max, which is why numeric.py
    // needed this at all) but named to match the Python/Lua backends' call
    // sites 1:1.
    public static double FMin(double a, double b) => Math.Min(a, b);
    public static double FMax(double a, double b) => Math.Max(a, b);
    public static double FClamp(double lo, double hi, double x) => FMax(lo, FMin(hi, x));

    // ECMAScript's ToInt32 abstract operation: NaN/+-Infinity -> 0, else
    // truncate toward zero then wrap into signed 32-bit range. `double`
    // arithmetic (not a `long` cast) handles the truncated magnitude even
    // when it exceeds Int64's range (e.g. 1e20): `t % 4294967296.0` is exact
    // (both operands, and the mathematical result, are exactly representable
    // as `double` — the result's magnitude is always < 2^32 < 2^53) whether
    // or not `t` itself fits in a `long`.
    public static int ToInt32(double x)
    {
        if (double.IsNaN(x) || double.IsInfinity(x))
        {
            return 0;
        }
        double t = Math.Truncate(x);
        double m = t % 4294967296.0;
        if (m < 0)
        {
            m += 4294967296.0;
        }
        uint u = (uint)m;
        return unchecked((int)u);
    }

    public static double SafeDiv(double x, double y) => x / y;
    public static double SafeFMod(double x, double y) => x % y;
    public static double SafeFloor(double x) => Math.Floor(x);
    public static double SafeCeil(double x) => Math.Ceiling(x);
    public static double SafeTrunc(double x) => Math.Truncate(x);

    // Matches JS's Math.round: round-half-up in MAGNITUDE (sign *
    // floor(|x| + 0.5)), NOT .NET's Math.Round (banker's rounding by
    // default) or plain floor semantics — the one Numeric.* wrapper besides
    // ToInt32/SafePow that needs real logic, not just a passthrough.
    public static double SafeRound(double x)
    {
        if (double.IsNaN(x) || double.IsInfinity(x))
        {
            return x;
        }
        double sign = x >= 0 ? 1.0 : -1.0;
        return sign * Math.Floor(Math.Abs(x) + 0.5);
    }

    public static double SafeSqrt(double x) => Math.Sqrt(x);
    public static double SafeLog(double x) => Math.Log(x);
    public static double SafeLog2(double x) => Math.Log2(x);
    public static double SafeLog10(double x) => Math.Log10(x);
    public static double SafeAsin(double x) => Math.Asin(x);
    public static double SafeAcos(double x) => Math.Acos(x);
    public static double SafeSin(double x) => Math.Sin(x);
    public static double SafeCos(double x) => Math.Cos(x);
    public static double SafeTan(double x) => Math.Tan(x);

    private static bool IsIntValued(double x) => !double.IsInfinity(x) && x == Math.Floor(x);
    private static bool IsOddInt(double x) => IsIntValued(x) && ((long)x % 2 != 0);

    // Matches ECMAScript's Number::exponentiate exactly (transcribed from
    // numeric.py's safe_pow, itself transcribed from the same spec) — NOT
    // .NET's native Math.Pow, which is IEEE-754/C99 `pow` and genuinely
    // disagrees with the JS spec at a couple of edge cases the spec
    // deliberately special-cases (e.g. `pow(+-1, +-Infinity)` is NaN per
    // ECMA-262 but 1 per C99/.NET's Math.Pow). Every special case below
    // mirrors safe_pow's own branch order 1:1; only the final fallback
    // (`Math.Pow(a, b)`) differs from Python's version, which needed a
    // try/except AND a `complex`-result check — neither applies in C#:
    // Math.Pow never throws, and (unlike Python's `**`) never silently
    // returns a complex number for a negative base with fractional exponent
    // (it returns NaN directly, matching JS already), so the a<0-and-
    // non-integer-exponent case is fully handled by the explicit guard
    // above the fallback, not by inspecting the fallback's result type.
    public static double SafePow(double a, double b)
    {
        if (double.IsNaN(a) || double.IsNaN(b))
        {
            return double.NaN;
        }
        if (b == 0.0)
        {
            return 1.0;
        }
        bool aNegZero = a == 0.0 && double.IsNegative(a);
        if (a == 0.0)
        {
            if (b < 0)
            {
                return (aNegZero && IsOddInt(b)) ? double.NegativeInfinity : double.PositiveInfinity;
            }
            return (aNegZero && IsOddInt(b)) ? -0.0 : 0.0;
        }
        if (double.IsInfinity(a))
        {
            if (b < 0)
            {
                return 0.0;
            }
            if (a < 0)
            {
                return IsOddInt(b) ? double.NegativeInfinity : double.PositiveInfinity;
            }
            return double.PositiveInfinity;
        }
        if (double.IsInfinity(b))
        {
            double aa = Math.Abs(a);
            if (aa == 1.0)
            {
                return double.NaN;
            }
            return (aa > 1.0) == (b > 0) ? double.PositiveInfinity : 0.0;
        }
        if (a < 0 && !IsIntValued(b))
        {
            return double.NaN;
        }
        return Math.Pow(a, b);
    }
}

File: packages/runtime-py/src/py/gltfi_runtime/numeric.py
# Numeric primitives every other module in this package leans on to match
# ECMAScript/Lua-backend semantics exactly, despite Python's math library
# (and its builtins) disagreeing with JS in several load-bearing ways. See
# the task report for the full list; summary of what's handled here:
#
#  - Python's `min`/`max` builtins are NOT NaN-propagating like JS's
#    Math.min/Math.max: `min(1, nan) == 1` but `min(nan, 1) == nan` (order-
#    dependent, since CPython's implementation just does a plain `<`/`>`
#    comparison against the running best-so-far, and every comparison
#    against NaN is False). JS's Math.min/Math.max always return NaN if
#    EITHER argument is NaN, order-independent. fmin/fmax below fix this.
#  - `x / y` for Python floats raises ZeroDivisionError when y == 0 instead
#    of returning +-Infinity/NaN like JS/Lua. safe_div fixes this, matching
#    IEEE-754 (and JS's) sign-of-zero-aware semantics.
#  - `math.fmod` raises ValueError ("math domain error") when the divisor is
#    0 or the dividend is infinite, instead of returning NaN like C's fmod
#    (which is what JS's `%` operator is specified to match). safe_fmod
#    fixes this.
#  - Python's floor division `//` always rounds toward -infinity (for BOTH
#    operands' signs); KHR_interactivity's int division truncates toward
#    zero (matches JS's `(a / b) | 0` and Lua's `math.floor`-then-correct
#    approach) — handled by to_int32 always being fed a value already
#    produced via true division (`/`, never `//`), not by this module
#    directly, but noted here since it's the same family of gotcha.
#  - `math.trunc`/`math.floor`/`math.ceil`/`round` all raise (ValueError for
#    NaN, OverflowError for +-Infinity) instead of returning NaN/Infinity
#    like JS's Math.floor/Math.ceil/Math.round. safe_floor/safe_ceil/
#    safe_round fix this (used by m.py's `floor`/`ceil`/`round` ops).
#  - `math.sqrt`/`math.log`/`math.log2`/`math.log10`/`math.asin`/`math.acos`
#    all raise ValueError ("math domain error") for out-of-domain FINITE
#    inputs (negative sqrt/log, |x|>1 for asin/acos) instead of returning
#    NaN like JS's Math.sqrt/Math.log/Math.asin/Math.acos. safe_sqrt/
#    safe_log/safe_log2/safe_log10/safe_asin/safe_acos fix this.
#  - `math.sin`/`math.cos`/`math.tan` raise ValueError for +-Infinity input
#    (NaN input is fine, propagates through natively) instead of returning
#    NaN like JS's Math.sin/Math.cos/Math.tan. safe_sin/safe_cos/safe_tan
#    fix this.
#  - The `**` operator, given a negative float base and a non-integer float
#    exponent, does NOT raise and does NOT return NaN: it silently returns a
#    Python `complex` number (e.g. `(-8.0) ** (1/3)` is a complex value with
#    a nonzero imaginary part) — a silent-wrong-type bug magnet, unlike JS's
#    Math.pow(-8, 1/3), which is NaN. safe_pow fixes this (also handles the
#    0**negative ZeroDivisionError and the sign-of-zero edge cases JS's
#    Number::exponentiate spec covers).
#  - Python ints are arbitrary-precision (bignum): every int-typed
#    KHR_interactivity arithmetic result must be wrapped back into signed
#    int32 range after the underlying operation — to_int32 (spec formula:
#    `((x & 0xFFFFFFFF) ^ 0x80000000) - 0x80000000`, applied to
#    `math.trunc(x)`, with NaN/+-Infinity inputs mapped to 0 first, mirroring
#    ECMAScript's ToInt32 abstract operation exactly).
import math

NAN = float("nan")
INF = math.inf
NEG_INF = -math.inf


def is_nan(x: float) -> bool:
    return x != x


def fmin(a: float, b: float) -> float:
    if a != a or b != b:
        return NAN
    return a if a < b else b


def fmax(a: float, b: float) -> float:
    if a != a or b != b:
        return NAN
    return a if a > b else b


def fclamp(lo: float, hi: float, x: float) -> float:
    return fmax(lo, fmin(hi, x))


def to_int32(x: float) -> int:
    if x != x or math.isinf(x):
        return 0
    n = math.trunc(x)
    return ((n & 0xFFFFFFFF) ^ 0x80000000) - 0x80000000


def safe_div(x: float, y: float) -> float:
    if y != 0.0:
        return x / y
    if x != x or x == 0.0:
        return NAN
    sign = math.copysign(1.0, x) * math.copysign(1.0, y)
    return INF if sign > 0 else NEG_INF


def safe_fmod(x: float, y: float) -> float:
    if y == 0.0 or x != x or y != y or math.isinf(x):
        return NAN
    if math.isinf(y):
        return x
    return math.fmod(x, y)


def safe_floor(x: float) -> float:
    if x != x or math.isinf(x):
        return x
    return float(math.floor(x))


def safe_ceil(x: float) -> float:
    if x != x or math.isinf(x):
        return x
    return float(math.ceil(x))


def safe_trunc(x: float) -> float:
    if x != x or math.isinf(x):
        return x
    return float(math.trunc(x))


def safe_round(x: float) -> float:
    # Matches JS's Math.round: round-half-up in magnitude (sign * floor(|x| + 0.5)),
    # NOT Python's builtin round() (banker's rounding) or math.floor semantics.
    if x != x or math.isinf(x):
        return x
    sign = 1.0 if x >= 0 else -1.0
    return sign * math.floor(abs(x) + 0.5)


def safe_sqrt(x: float) -> float:
    if x != x:
        return NAN
    if x < 0:
        return NAN
    return math.sqrt(x)


def safe_log(x: float) -> float:
    if x != x:
        return NAN
    if x < 0:
        return NAN
    if x == 0:
        return NEG_INF
    return math.log(x)


def safe_log2(x: float) -> float:
    if x != x:
        return NAN
    if x < 0:
        return NAN
    if x == 0:
        return NEG_INF
    return math.log2(x)


def safe_log10(x: float) -> float:
    if x != x:
        return NAN
    if x < 0:
        return NAN
    if x == 0:
        return NEG_INF
    return math.log10(x)


def safe_asin(x: float) -> float:
    if x != x or x < -1 or x > 1:
        return NAN
    return math.asin(x)


def safe_acos(x: float) -> float:
    if x != x or x < -1 or x > 1:
        return NAN
    return math.acos(x)


def safe_sin(x: float) -> float:
    if x != x:
        return NAN
    if math.isinf(x):
        return NAN
    return math.sin(x)


def safe_cos(x: float) -> float:
    if x != x:
        return NAN
    if math.isinf(x):
        return NAN
    return math.cos(x)


def safe_tan(x: float) -> float:
    if x != x:
        return NAN
    if math.isinf(x):
        return NAN
    return math.tan(x)


def _is_int_valued(x: float) -> bool:
    return not math.isinf(x) and x == math.floor(x)


def _is_odd_int(x: float) -> bool:
    return _is_int_valued(x) and (int(x) % 2 != 0)


def safe_pow(a: float, b: float) -> float:
    if a != a or b != b:
        return NAN
    if b == 0.0:
        return 1.0
    a_neg_zero = a == 0.0 and math.copysign(1.0, a) < 0
    if a == 0.0:
        if b < 0:
            return NEG_INF if (a_neg_zero and _is_odd_int(b)) else INF
        return -0.0 if (a_neg_zero and _is_odd_int(b)) else 0.0
    if math.isinf(a):
        if b < 0:
            return 0.0
        if a < 0:
            return NEG_INF if _is_odd_int(b) else INF
        return INF
    if math.isinf(b):
        aa = abs(a)
        if aa == 1.0:
            return NAN
        return INF if (aa > 1.0) == (b > 0) else 0.0
    if a < 0 and not _is_int_valued(b):
        return NAN
    try:
        result = a ** b
    except (OverflowError, ZeroDivisionError):
        return INF
    if isinstance(result, complex):
        return NAN
    return float(result)
